Error message for type mismatch between impl Traits needs better explanation

The message "expected type `impl Trait` (opaque type), found type `impl Trait` (opaque type)" sounds like a paradox: the compiler has found exactly what it expected, and yet it doesn't match.

The explanatory text of E0308 is the same as for all other banal type errors, so it's not addressing this problem specifically.

From the questions I'm seeing in the user forum I have an impression that users think `impl Trait` is like `dyn Trait` and try to use it as an abstraction for mixing different types.

```rust
fn f1() -> impl Copy {
    "1"
}
fn f2() -> impl Copy {
    2
}

fn main() {
    let mut x = f1();
    x = f2();
}
```

> 10 |     x = f2();
   |         ^^^^ expected opaque type, found a different opaque type
   |
   = note: expected type `impl std::marker::Copy` (opaque type)
              found type `impl std::marker::Copy` (opaque type)

----

Suggestions:

* Make it a separate error, so that it'll get its own `rustc --explain` document.

* If possible, instead of "(opaque type)" print more internal type information, e.g. like for closures:

      = note: expected type `impl [std::marker::Copy@src/main.rs:2:10]`
              found type `impl [std::marker::Copy@src/main.rs:5:10]`

      = note: each `impl Trait`, even for the same trait, is a different type.
      = help: consider using `Box<dyn Trait>` or separate generic parameters.

Current output:

```
error[E0308]: mismatched types
  --> src/main.rs:10:9
   |
10 |     x = f2();
   |         ^^^^ expected opaque type, found a different opaque type
   |
   = note: expected type `impl std::marker::Copy` (opaque type at <src/main.rs:1:12>)
              found type `impl std::marker::Copy` (opaque type at <src/main.rs:4:12>)
   = note: distinct uses of `impl Trait` result in different opaque types
```

The suggestion to use `Box<dyn Trait>` could be interesting as text, but giving a structured suggestion would be quite difficult.

--- a/src/librustc/infer/error_reporting/mod.rs
+++ b/src/librustc/infer/error_reporting/mod.rs
@@ -2039,7 +2039,7 @@ impl<'tcx> ObligationCause<'tcx> {
 
 /// This is a bare signal of what kind of type we're dealing with. `ty::TyKind` tracks
 /// extra information about each type, but we only care about the category.
-#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
 crate enum TyCategory {
     Closure,
     Opaque,

--- a/src/librustc/infer/error_reporting/mod.rs
+++ b/src/librustc/infer/error_reporting/mod.rs
@@ -1303,10 +1303,39 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             _ => {}
         }
 
+        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+        enum TyKind {
+            Closure,
+            Opaque,
+            Generator,
+            Foreign,
+        }
+
+        impl TyKind {
+            fn descr(&self) -> &'static str {
+                match self {
+                    Self::Closure => "closure",
+                    Self::Opaque => "opaque type",
+                    Self::Generator => "generator",
+                    Self::Foreign => "foreign type",
+                }
+            }
+
+            fn from_ty(ty: Ty<'_>) -> Option<(Self, DefId)> {
+                match ty.kind {
+                    ty::Closure(def_id, _) => Some((Self::Closure, def_id)),
+                    ty::Opaque(def_id, _) => Some((Self::Opaque, def_id)),
+                    ty::Generator(def_id, ..) => Some((Self::Generator, def_id)),
+                    ty::Foreign(def_id) => Some((Self::Foreign, def_id)),
+                    _ => None,
+                }
+            }
+        }
+
         struct OpaqueTypesVisitor<'tcx> {
-            types: FxHashMap<&'static str, FxHashSet<Span>>,
-            expected: FxHashMap<&'static str, FxHashSet<Span>>,
-            found: FxHashMap<&'static str, FxHashSet<Span>>,
+            types: FxHashMap<TyKind, FxHashSet<Span>>,
+            expected: FxHashMap<TyKind, FxHashSet<Span>>,
+            found: FxHashMap<TyKind, FxHashSet<Span>>,
             ignore_span: Span,
             tcx: TyCtxt<'tcx>,
         }
@@ -1350,7 +1379,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                                     },
                                     if count > 1 { "one of the " } else { "" },
                                     target,
-                                    key,
+                                    key.descr(),
                                     pluralize!(count),
                                 ),
                             );
@@ -1362,18 +1391,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
 
         impl<'tcx> ty::fold::TypeVisitor<'tcx> for OpaqueTypesVisitor<'tcx> {
             fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
-                let kind = match t.kind {
-                    ty::Closure(..) => "closure",
-                    ty::Opaque(..) => "opaque type",
-                    ty::Generator(..) => "generator",
-                    ty::Foreign(..) => "foreign type",
-                    _ => "",
-                };
-                if let ty::Closure(def_id, _)
-                | ty::Opaque(def_id, _)
-                | ty::Generator(def_id, ..)
-                | ty::Foreign(def_id) = t.kind
-                {
+                if let Some((kind, def_id)) = TyKind::from_ty(t) {
                     let span = self.tcx.def_span(def_id);
                     // Avoid cluttering the output when the "found" and error span overlap:
                     //

--- a/src/librustc/infer/error_reporting/mod.rs
+++ b/src/librustc/infer/error_reporting/mod.rs
@@ -1339,16 +1339,16 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                             err.span_label(
                                 *sp,
                                 format!(
-                                    "{}this is {}the {} {}{}",
-                                    if sp.is_desugaring(DesugaringKind::Async) {
-                                        "in the desugared `async fn`, "
-                                    } else {
-                                        ""
-                                    },
-                                    if count > 1 { "one of" } else { "" },
+                                    "{}the {} {}{}{}",
+                                    if count > 1 { "one of " } else { "" },
                                     target,
                                     key,
                                     pluralize!(count),
+                                    if sp.is_desugaring(DesugaringKind::Async) {
+                                        " in the `Output` of this `async fn`"
+                                    } else {
+                                        ""
+                                    },
                                 ),
                             );
                         }
@@ -1364,18 +1364,24 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                     ty::Opaque(..) => "opaque type",
                     _ => "",
                 };
-                match t.kind {
-                    ty::Closure(def_id, _) | ty::Opaque(def_id, _) => {
-                        let span = self.tcx.def_span(def_id);
-                        debug!("note_type_err visit_ty {:?}", span.macro_backtrace());
-                        if !self.ignore_span.overlaps(span)
-                            && !self.expected.values().any(|exp| exp.iter().any(|sp| *sp == span))
-                        {
-                            let entry = self.types.entry(kind).or_default();
-                            entry.insert(span);
-                        }
+                if let ty::Closure(def_id, _) | ty::Opaque(def_id, _) = t.kind {
+                    let span = self.tcx.def_span(def_id);
+                    // Avoid cluttering the output when the "found" and error span overlap:
+                    //
+                    // error[E0308]: mismatched types
+                    //   --> $DIR/issue-20862.rs:2:5
+                    //    |
+                    // LL |     |y| x + y
+                    //    |     ^^^^^^^^^
+                    //    |     |
+                    //    |     the found closure
+                    //    |     expected `()`, found closure
+                    //    |
+                    //    = note: expected unit type `()`
+                    //                 found closure `[closure@$DIR/issue-20862.rs:2:5: 2:14 x:_]`
+                    if !self.ignore_span.overlaps(span) {
+                        self.types.entry(kind).or_default().insert(span);
                     }
-                    _ => {}
                 }
                 t.super_visit_with(self)
             }

--- a/src/librustc/infer/error_reporting/mod.rs
+++ b/src/librustc/infer/error_reporting/mod.rs
@@ -1340,16 +1340,18 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                             err.span_label(
                                 *sp,
                                 format!(
-                                    "{}the {} {}{}{}",
-                                    if count > 1 { "one of " } else { "" },
-                                    target,
-                                    key,
-                                    pluralize!(count),
+                                    "{}{}{} {}{}",
                                     if sp.is_desugaring(DesugaringKind::Async) {
-                                        " in the `Output` of this `async fn`"
+                                        "the `Output` of this `async fn`'s "
+                                    } else if count == 1 {
+                                        "the "
                                     } else {
                                         ""
                                     },
+                                    if count > 1 { "one of the " } else { "" },
+                                    target,
+                                    key,
+                                    pluralize!(count),
                                 ),
                             );
                         }

--- a/src/librustc/infer/error_reporting/mod.rs
+++ b/src/librustc/infer/error_reporting/mod.rs
@@ -71,6 +71,7 @@ use rustc_hir::Node;
 use errors::{
     pluralize, struct_span_err, Applicability, DiagnosticBuilder, DiagnosticStyledString,
 };
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
 use rustc_error_codes::*;
 use rustc_span::{DesugaringKind, Pos, Span};
 use rustc_target::spec::abi;
@@ -1362,9 +1363,15 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 let kind = match t.kind {
                     ty::Closure(..) => "closure",
                     ty::Opaque(..) => "opaque type",
+                    ty::Generator(..) => "generator",
+                    ty::Foreign(..) => "foreign type",
                     _ => "",
                 };
-                if let ty::Closure(def_id, _) | ty::Opaque(def_id, _) = t.kind {
+                if let ty::Closure(def_id, _)
+                | ty::Opaque(def_id, _)
+                | ty::Generator(def_id, ..)
+                | ty::Foreign(def_id) = t.kind
+                {
                     let span = self.tcx.def_span(def_id);
                     // Avoid cluttering the output when the "found" and error span overlap:
                     //

--- a/src/librustc/infer/error_reporting/mod.rs
+++ b/src/librustc/infer/error_reporting/mod.rs
@@ -68,9 +68,11 @@ use rustc_hir as hir;
 use rustc_hir::def_id::DefId;
 use rustc_hir::Node;
 
-use errors::{struct_span_err, Applicability, DiagnosticBuilder, DiagnosticStyledString};
+use errors::{
+    pluralize, struct_span_err, Applicability, DiagnosticBuilder, DiagnosticStyledString,
+};
 use rustc_error_codes::*;
-use rustc_span::{Pos, Span};
+use rustc_span::{DesugaringKind, Pos, Span};
 use rustc_target::spec::abi;
 use std::{cmp, fmt};
 
@@ -1289,6 +1291,8 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         mut values: Option<ValuePairs<'tcx>>,
         terr: &TypeError<'tcx>,
     ) {
+        let span = cause.span(self.tcx);
+
         // For some types of errors, expected-found does not make
         // sense, so just ignore the values we were given.
         match terr {
@@ -1296,6 +1300,85 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 values = None;
             }
             _ => {}
+        }
+
+        struct OpaqueTypesVisitor<'tcx> {
+            types: FxHashMap<&'static str, FxHashSet<Span>>,
+            expected: FxHashMap<&'static str, FxHashSet<Span>>,
+            found: FxHashMap<&'static str, FxHashSet<Span>>,
+            ignore_span: Span,
+            tcx: TyCtxt<'tcx>,
+        }
+
+        impl<'tcx> OpaqueTypesVisitor<'tcx> {
+            fn visit_expected_found(
+                tcx: TyCtxt<'tcx>,
+                expected: Ty<'tcx>,
+                found: Ty<'tcx>,
+                ignore_span: Span,
+            ) -> Self {
+                let mut types_visitor = OpaqueTypesVisitor {
+                    types: Default::default(),
+                    expected: Default::default(),
+                    found: Default::default(),
+                    ignore_span,
+                    tcx,
+                };
+                expected.visit_with(&mut types_visitor);
+                std::mem::swap(&mut types_visitor.expected, &mut types_visitor.types);
+                found.visit_with(&mut types_visitor);
+                std::mem::swap(&mut types_visitor.found, &mut types_visitor.types);
+                types_visitor
+            }
+
+            fn report(&self, err: &mut DiagnosticBuilder<'_>) {
+                for (target, types) in &[("expected", &self.expected), ("found", &self.found)] {
+                    for (key, values) in types.iter() {
+                        let count = values.len();
+                        for sp in values {
+                            err.span_label(
+                                *sp,
+                                format!(
+                                    "{}this is {}the {} {}{}",
+                                    if sp.is_desugaring(DesugaringKind::Async) {
+                                        "in the desugared `async fn`, "
+                                    } else {
+                                        ""
+                                    },
+                                    if count > 1 { "one of" } else { "" },
+                                    target,
+                                    key,
+                                    pluralize!(count),
+                                ),
+                            );
+                        }
+                    }
+                }
+            }
+        }
+
+        impl<'tcx> ty::fold::TypeVisitor<'tcx> for OpaqueTypesVisitor<'tcx> {
+            fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
+                let kind = match t.kind {
+                    ty::Closure(..) => "closure",
+                    ty::Opaque(..) => "opaque type",
+                    _ => "",
+                };
+                match t.kind {
+                    ty::Closure(def_id, _) | ty::Opaque(def_id, _) => {
+                        let span = self.tcx.def_span(def_id);
+                        debug!("note_type_err visit_ty {:?}", span.macro_backtrace());
+                        if !self.ignore_span.overlaps(span)
+                            && !self.expected.values().any(|exp| exp.iter().any(|sp| *sp == span))
+                        {
+                            let entry = self.types.entry(kind).or_default();
+                            entry.insert(span);
+                        }
+                    }
+                    _ => {}
+                }
+                t.super_visit_with(self)
+            }
         }
 
         debug!("note_type_err(diag={:?})", diag);
@@ -1306,6 +1389,13 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                     ValuePairs::Types(exp_found) => {
                         let is_simple_err =
                             exp_found.expected.is_simple_text() && exp_found.found.is_simple_text();
+                        OpaqueTypesVisitor::visit_expected_found(
+                            self.tcx,
+                            exp_found.expected,
+                            exp_found.found,
+                            span,
+                        )
+                        .report(diag);
 
                         (is_simple_err, Some(exp_found))
                     }
@@ -1323,8 +1413,6 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             }
         };
 
-        let span = cause.span(self.tcx);
-
         // Ignore msg for object safe coercion
         // since E0038 message will be printed
         match terr {
@@ -1336,7 +1424,6 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 }
             }
         };
-
         if let Some((expected, found)) = expected_found {
             let expected_label = exp_found.map_or("type".into(), |ef| ef.expected.prefix_string());
             let found_label = exp_found.map_or("type".into(), |ef| ef.found.prefix_string());

--- a/src/librustc/infer/error_reporting/mod.rs
+++ b/src/librustc/infer/error_reporting/mod.rs
@@ -1303,39 +1303,8 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             _ => {}
         }
 
-        /// This is a bare signal of what kind of type we're dealing with. `ty::TyKind` tracks
-        /// extra information about each type, but we only care about the category.
-        #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
-        enum TyCategory {
-            Closure,
-            Opaque,
-            Generator,
-            Foreign,
-        }
-
-        impl TyCategory {
-            fn descr(&self) -> &'static str {
-                match self {
-                    Self::Closure => "closure",
-                    Self::Opaque => "opaque type",
-                    Self::Generator => "generator",
-                    Self::Foreign => "foreign type",
-                }
-            }
-
-            fn from_ty(ty: Ty<'_>) -> Option<(Self, DefId)> {
-                match ty.kind {
-                    ty::Closure(def_id, _) => Some((Self::Closure, def_id)),
-                    ty::Opaque(def_id, _) => Some((Self::Opaque, def_id)),
-                    ty::Generator(def_id, ..) => Some((Self::Generator, def_id)),
-                    ty::Foreign(def_id) => Some((Self::Foreign, def_id)),
-                    _ => None,
-                }
-            }
-        }
-
         struct OpaqueTypesVisitor<'tcx> {
-            types: FxHashMap<TyKind, FxHashSet<Span>>,
+            types: FxHashMap<TyCategory, FxHashSet<Span>>,
             expected: FxHashMap<TyCategory, FxHashSet<Span>>,
             found: FxHashMap<TyCategory, FxHashSet<Span>>,
             ignore_span: Span,
@@ -1375,7 +1344,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 &self,
                 err: &mut DiagnosticBuilder<'_>,
                 target: &str,
-                types: &FxHashMap<TyKind, FxHashSet<Span>>,
+                types: &FxHashMap<TyCategory, FxHashSet<Span>>,
             ) {
                 for (key, values) in types.iter() {
                     let count = values.len();
@@ -1394,7 +1363,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                                 },
                                 if count > 1 { "one of the " } else { "" },
                                 target,
-                                key,
+                                kind,
                                 pluralize!(count),
                             ),
                         );
@@ -1405,7 +1374,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
 
         impl<'tcx> ty::fold::TypeVisitor<'tcx> for OpaqueTypesVisitor<'tcx> {
             fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
-                if let Some((kind, def_id)) = TyKind::from_ty(t) {
+                if let Some((kind, def_id)) = TyCategory::from_ty(t) {
                     let span = self.tcx.def_span(def_id);
                     // Avoid cluttering the output when the "found" and error span overlap:
                     //
@@ -2064,6 +2033,37 @@ impl<'tcx> ObligationCause<'tcx> {
             IntrinsicType => "intrinsic has the correct type",
             MethodReceiver => "method receiver has the correct type",
             _ => "types are compatible",
+        }
+    }
+}
+
+/// This is a bare signal of what kind of type we're dealing with. `ty::TyKind` tracks
+/// extra information about each type, but we only care about the category.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+crate enum TyCategory {
+    Closure,
+    Opaque,
+    Generator,
+    Foreign,
+}
+
+impl TyCategory {
+    fn descr(&self) -> &'static str {
+        match self {
+            Self::Closure => "closure",
+            Self::Opaque => "opaque type",
+            Self::Generator => "generator",
+            Self::Foreign => "foreign type",
+        }
+    }
+
+    pub fn from_ty(ty: Ty<'_>) -> Option<(Self, DefId)> {
+        match ty.kind {
+            ty::Closure(def_id, _) => Some((Self::Closure, def_id)),
+            ty::Opaque(def_id, _) => Some((Self::Opaque, def_id)),
+            ty::Generator(def_id, ..) => Some((Self::Generator, def_id)),
+            ty::Foreign(def_id) => Some((Self::Foreign, def_id)),
+            _ => None,
         }
     }
 }

--- a/src/librustc/infer/error_reporting/mod.rs
+++ b/src/librustc/infer/error_reporting/mod.rs
@@ -1303,15 +1303,17 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             _ => {}
         }
 
+        /// This is a bare signal of what kind of type we're dealing with. `ty::TyKind` tracks
+        /// extra information about each type, but we only care about the category.
         #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
-        enum TyKind {
+        enum TyCategory {
             Closure,
             Opaque,
             Generator,
             Foreign,
         }
 
-        impl TyKind {
+        impl TyCategory {
             fn descr(&self) -> &'static str {
                 match self {
                     Self::Closure => "closure",
@@ -1334,8 +1336,8 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
 
         struct OpaqueTypesVisitor<'tcx> {
             types: FxHashMap<TyKind, FxHashSet<Span>>,
-            expected: FxHashMap<TyKind, FxHashSet<Span>>,
-            found: FxHashMap<TyKind, FxHashSet<Span>>,
+            expected: FxHashMap<TyCategory, FxHashSet<Span>>,
+            found: FxHashMap<TyCategory, FxHashSet<Span>>,
             ignore_span: Span,
             tcx: TyCtxt<'tcx>,
         }
@@ -1354,6 +1356,9 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                     ignore_span,
                     tcx,
                 };
+                // The visitor puts all the relevant encountered types in `self.types`, but in
+                // here we want to visit two separate types with no relation to each other, so we
+                // move the results from `types` to `expected` or `found` as appropriate.
                 expected.visit_with(&mut types_visitor);
                 std::mem::swap(&mut types_visitor.expected, &mut types_visitor.types);
                 found.visit_with(&mut types_visitor);
@@ -1362,28 +1367,37 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             }
 
             fn report(&self, err: &mut DiagnosticBuilder<'_>) {
-                for (target, types) in &[("expected", &self.expected), ("found", &self.found)] {
-                    for (key, values) in types.iter() {
-                        let count = values.len();
-                        for sp in values {
-                            err.span_label(
-                                *sp,
-                                format!(
-                                    "{}{}{} {}{}",
-                                    if sp.is_desugaring(DesugaringKind::Async) {
-                                        "the `Output` of this `async fn`'s "
-                                    } else if count == 1 {
-                                        "the "
-                                    } else {
-                                        ""
-                                    },
-                                    if count > 1 { "one of the " } else { "" },
-                                    target,
-                                    key.descr(),
-                                    pluralize!(count),
-                                ),
-                            );
-                        }
+                self.add_labels_for_types(err, "expected", &self.expected);
+                self.add_labels_for_types(err, "found", &self.found);
+            }
+
+            fn add_labels_for_types(
+                &self,
+                err: &mut DiagnosticBuilder<'_>,
+                target: &str,
+                types: &FxHashMap<TyKind, FxHashSet<Span>>,
+            ) {
+                for (key, values) in types.iter() {
+                    let count = values.len();
+                    let kind = key.descr();
+                    for sp in values {
+                        err.span_label(
+                            *sp,
+                            format!(
+                                "{}{}{} {}{}",
+                                if sp.is_desugaring(DesugaringKind::Async) {
+                                    "the `Output` of this `async fn`'s "
+                                } else if count == 1 {
+                                    "the "
+                                } else {
+                                    ""
+                                },
+                                if count > 1 { "one of the " } else { "" },
+                                target,
+                                key,
+                                pluralize!(count),
+                            ),
+                        );
                     }
                 }
             }

--- a/src/librustc/traits/error_reporting.rs
+++ b/src/librustc/traits/error_reporting.rs
@@ -6,7 +6,7 @@ use super::{
     TraitNotObjectSafe,
 };
 
-use crate::infer::error_reporting::TypeAnnotationNeeded as ErrorCode;
+use crate::infer::error_reporting::{TyCategory, TypeAnnotationNeeded as ErrorCode};
 use crate::infer::type_variable::{TypeVariableOrigin, TypeVariableOriginKind};
 use crate::infer::{self, InferCtxt};
 use crate::mir::interpret::ErrorHandled;
@@ -676,15 +676,8 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                     Some(t) => Some(t),
                     None => {
                         let ty = parent_trait_ref.skip_binder().self_ty();
-                        let span = if let ty::Closure(def_id, _)
-                        | ty::Opaque(def_id, _)
-                        | ty::Generator(def_id, ..)
-                        | ty::Foreign(def_id) = ty.kind
-                        {
-                            Some(self.tcx.def_span(def_id))
-                        } else {
-                            None
-                        };
+                        let span =
+                            TyCategory::from_ty(ty).map(|(_, def_id)| self.tcx.def_span(def_id));
                         Some((ty.to_string(), span))
                     }
                 }

--- a/src/librustc/traits/error_reporting.rs
+++ b/src/librustc/traits/error_reporting.rs
@@ -446,7 +446,7 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 flags.push((sym::from_method, Some(method.to_string())));
             }
         }
-        if let Some(t) = self.get_parent_trait_ref(&obligation.cause.code) {
+        if let Some((t, _)) = self.get_parent_trait_ref(&obligation.cause.code) {
             flags.push((sym::parent_trait, Some(t)));
         }
 
@@ -665,13 +665,28 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
     }
 
     /// Gets the parent trait chain start
-    fn get_parent_trait_ref(&self, code: &ObligationCauseCode<'tcx>) -> Option<String> {
+    fn get_parent_trait_ref(
+        &self,
+        code: &ObligationCauseCode<'tcx>,
+    ) -> Option<(String, Option<Span>)> {
         match code {
             &ObligationCauseCode::BuiltinDerivedObligation(ref data) => {
                 let parent_trait_ref = self.resolve_vars_if_possible(&data.parent_trait_ref);
                 match self.get_parent_trait_ref(&data.parent_code) {
                     Some(t) => Some(t),
-                    None => Some(parent_trait_ref.skip_binder().self_ty().to_string()),
+                    None => {
+                        let ty = parent_trait_ref.skip_binder().self_ty();
+                        let span = if let ty::Closure(def_id, _)
+                        | ty::Opaque(def_id, _)
+                        | ty::Generator(def_id, ..)
+                        | ty::Foreign(def_id) = ty.kind
+                        {
+                            Some(self.tcx.def_span(def_id))
+                        } else {
+                            None
+                        };
+                        Some((ty.to_string(), span))
+                    }
                 }
             }
             _ => None,
@@ -719,9 +734,15 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                             return;
                         }
                         let trait_ref = trait_predicate.to_poly_trait_ref();
-                        let (post_message, pre_message) = self
+                        let (post_message, pre_message, type_def) = self
                             .get_parent_trait_ref(&obligation.cause.code)
-                            .map(|t| (format!(" in `{}`", t), format!("within `{}`, ", t)))
+                            .map(|(t, s)| {
+                                (
+                                    format!(" in `{}`", t),
+                                    format!("within `{}`, ", t),
+                                    s.map(|s| (format!("within this `{}`", t), s)),
+                                )
+                            })
                             .unwrap_or_default();
 
                         let OnUnimplementedNote { message, label, note, enclosing_scope } =
@@ -794,6 +815,9 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                             err.help(&explanation);
                         } else {
                             err.span_label(span, explanation);
+                        }
+                        if let Some((msg, span)) = type_def {
+                            err.span_label(span, &msg);
                         }
                         if let Some(ref s) = note {
                             // If it has a custom `#[rustc_on_unimplemented]` note, let's display it

--- a/src/librustc_typeck/check/mod.rs
+++ b/src/librustc_typeck/check/mod.rs
@@ -4743,14 +4743,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         .join(", ");
                 }
                 Some(Node::Expr(hir::Expr {
-                    kind: ExprKind::Closure(_, _, body_id, closure_span, _),
+                    kind: ExprKind::Closure(_, _, body_id, _, _),
                     span: full_closure_span,
                     ..
                 })) => {
                     if *full_closure_span == expr.span {
                         return false;
                     }
-                    err.span_label(*closure_span, "closure defined here");
                     msg = "call this closure";
                     let body = hir.body(*body_id);
                     sugg_call = body

--- a/src/test/ui/async-await/dont-suggest-missing-await.stderr
+++ b/src/test/ui/async-await/dont-suggest-missing-await.stderr
@@ -1,6 +1,9 @@
 error[E0308]: mismatched types
   --> $DIR/dont-suggest-missing-await.rs:14:18
    |
+LL | async fn make_u32() -> u32 {
+   |                        --- in the desugared `async fn`, this is the found opaque type
+...
 LL |         take_u32(x)
    |                  ^ expected `u32`, found opaque type
    |

--- a/src/test/ui/async-await/dont-suggest-missing-await.stderr
+++ b/src/test/ui/async-await/dont-suggest-missing-await.stderr
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
   --> $DIR/dont-suggest-missing-await.rs:14:18
    |
 LL | async fn make_u32() -> u32 {
-   |                        --- the found opaque type in the `Output` of this `async fn`
+   |                        --- the `Output` of this `async fn`'s found opaque type
 ...
 LL |         take_u32(x)
    |                  ^ expected `u32`, found opaque type

--- a/src/test/ui/async-await/dont-suggest-missing-await.stderr
+++ b/src/test/ui/async-await/dont-suggest-missing-await.stderr
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
   --> $DIR/dont-suggest-missing-await.rs:14:18
    |
 LL | async fn make_u32() -> u32 {
-   |                        --- in the desugared `async fn`, this is the found opaque type
+   |                        --- the found opaque type in the `Output` of this `async fn`
 ...
 LL |         take_u32(x)
    |                  ^ expected `u32`, found opaque type

--- a/src/test/ui/async-await/issue-64130-3-other.stderr
+++ b/src/test/ui/async-await/issue-64130-3-other.stderr
@@ -3,6 +3,9 @@ error[E0277]: the trait bound `Foo: Qux` is not satisfied in `impl std::future::
    |
 LL | fn is_qux<T: Qux>(t: T) { }
    |    ------    --- required by this bound in `is_qux`
+LL | 
+LL | async fn bar() {
+   |                - within this `impl std::future::Future`
 ...
 LL |     is_qux(bar());
    |     ^^^^^^ within `impl std::future::Future`, the trait `Qux` is not implemented for `Foo`

--- a/src/test/ui/async-await/suggest-missing-await-closure.stderr
+++ b/src/test/ui/async-await/suggest-missing-await-closure.stderr
@@ -1,6 +1,9 @@
 error[E0308]: mismatched types
   --> $DIR/suggest-missing-await-closure.rs:16:18
    |
+LL | async fn make_u32() -> u32 {
+   |                        --- in the desugared `async fn`, this is the found opaque type
+...
 LL |         take_u32(x)
    |                  ^
    |                  |

--- a/src/test/ui/async-await/suggest-missing-await-closure.stderr
+++ b/src/test/ui/async-await/suggest-missing-await-closure.stderr
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
   --> $DIR/suggest-missing-await-closure.rs:16:18
    |
 LL | async fn make_u32() -> u32 {
-   |                        --- in the desugared `async fn`, this is the found opaque type
+   |                        --- the found opaque type in the `Output` of this `async fn`
 ...
 LL |         take_u32(x)
    |                  ^

--- a/src/test/ui/async-await/suggest-missing-await-closure.stderr
+++ b/src/test/ui/async-await/suggest-missing-await-closure.stderr
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
   --> $DIR/suggest-missing-await-closure.rs:16:18
    |
 LL | async fn make_u32() -> u32 {
-   |                        --- the found opaque type in the `Output` of this `async fn`
+   |                        --- the `Output` of this `async fn`'s found opaque type
 ...
 LL |         take_u32(x)
    |                  ^

--- a/src/test/ui/async-await/suggest-missing-await.stderr
+++ b/src/test/ui/async-await/suggest-missing-await.stderr
@@ -1,6 +1,9 @@
 error[E0308]: mismatched types
   --> $DIR/suggest-missing-await.rs:13:14
    |
+LL | async fn make_u32() -> u32 {
+   |                        --- in the desugared `async fn`, this is the found opaque type
+...
 LL |     take_u32(x)
    |              ^
    |              |

--- a/src/test/ui/async-await/suggest-missing-await.stderr
+++ b/src/test/ui/async-await/suggest-missing-await.stderr
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
   --> $DIR/suggest-missing-await.rs:13:14
    |
 LL | async fn make_u32() -> u32 {
-   |                        --- in the desugared `async fn`, this is the found opaque type
+   |                        --- the found opaque type in the `Output` of this `async fn`
 ...
 LL |     take_u32(x)
    |              ^
@@ -16,6 +16,9 @@ LL |     take_u32(x)
 error[E0308]: mismatched types
   --> $DIR/suggest-missing-await.rs:23:5
    |
+LL | async fn dummy() {}
+   |                  - the found opaque type in the `Output` of this `async fn`
+...
 LL |     dummy()
    |     ^^^^^^^ expected `()`, found opaque type
    |

--- a/src/test/ui/async-await/suggest-missing-await.stderr
+++ b/src/test/ui/async-await/suggest-missing-await.stderr
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
   --> $DIR/suggest-missing-await.rs:13:14
    |
 LL | async fn make_u32() -> u32 {
-   |                        --- the found opaque type in the `Output` of this `async fn`
+   |                        --- the `Output` of this `async fn`'s found opaque type
 ...
 LL |     take_u32(x)
    |              ^
@@ -17,7 +17,7 @@ error[E0308]: mismatched types
   --> $DIR/suggest-missing-await.rs:23:5
    |
 LL | async fn dummy() {}
-   |                  - the found opaque type in the `Output` of this `async fn`
+   |                  - the `Output` of this `async fn`'s found opaque type
 ...
 LL |     dummy()
    |     ^^^^^^^ expected `()`, found opaque type

--- a/src/test/ui/closures/closure-reform-bad.stderr
+++ b/src/test/ui/closures/closure-reform-bad.stderr
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
   --> $DIR/closure-reform-bad.rs:11:15
    |
 LL |     let f = |s: &str| println!("{}{}", s, string);
-   |             ------------------------------------- this is the found closure
+   |             ------------------------------------- the found closure
 LL |     call_bare(f)
    |               ^ expected fn pointer, found closure
    |

--- a/src/test/ui/closures/closure-reform-bad.stderr
+++ b/src/test/ui/closures/closure-reform-bad.stderr
@@ -1,6 +1,8 @@
 error[E0308]: mismatched types
   --> $DIR/closure-reform-bad.rs:11:15
    |
+LL |     let f = |s: &str| println!("{}{}", s, string);
+   |             ------------------------------------- this is the found closure
 LL |     call_bare(f)
    |               ^ expected fn pointer, found closure
    |

--- a/src/test/ui/extern/extern-types-distinct-types.stderr
+++ b/src/test/ui/extern/extern-types-distinct-types.stderr
@@ -1,6 +1,11 @@
 error[E0308]: mismatched types
   --> $DIR/extern-types-distinct-types.rs:9:5
    |
+LL |     type A;
+   |     ------- the found foreign type
+LL |     type B;
+   |     ------- the expected foreign type
+...
 LL |     r
    |     ^ expected extern type `B`, found extern type `A`
    |

--- a/src/test/ui/impl-trait/auto-trait-leak.stderr
+++ b/src/test/ui/impl-trait/auto-trait-leak.stderr
@@ -77,6 +77,9 @@ LL | fn send<T: Send>(_: T) {}
 ...
 LL |     send(cycle2().clone());
    |     ^^^^ `std::rc::Rc<std::string::String>` cannot be sent between threads safely
+...
+LL | fn cycle2() -> impl Clone {
+   |                ---------- within this `impl std::clone::Clone`
    |
    = help: within `impl std::clone::Clone`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<std::string::String>`
    = note: required because it appears within the type `impl std::clone::Clone`

--- a/src/test/ui/impl-trait/auto-trait-leak2.stderr
+++ b/src/test/ui/impl-trait/auto-trait-leak2.stderr
@@ -1,6 +1,9 @@
 error[E0277]: `std::rc::Rc<std::cell::Cell<i32>>` cannot be sent between threads safely
   --> $DIR/auto-trait-leak2.rs:13:5
    |
+LL | fn before() -> impl Fn(i32) {
+   |                ------------ within this `impl std::ops::Fn<(i32,)>`
+...
 LL | fn send<T: Send>(_: T) {}
    |    ----    ---- required by this bound in `send`
 ...
@@ -19,6 +22,9 @@ LL | fn send<T: Send>(_: T) {}
 ...
 LL |     send(after());
    |     ^^^^ `std::rc::Rc<std::cell::Cell<i32>>` cannot be sent between threads safely
+...
+LL | fn after() -> impl Fn(i32) {
+   |               ------------ within this `impl std::ops::Fn<(i32,)>`
    |
    = help: within `impl std::ops::Fn<(i32,)>`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<std::cell::Cell<i32>>`
    = note: required because it appears within the type `[closure@$DIR/auto-trait-leak2.rs:24:5: 24:22 p:std::rc::Rc<std::cell::Cell<i32>>]`

--- a/src/test/ui/impl-trait/equality2.stderr
+++ b/src/test/ui/impl-trait/equality2.stderr
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
   --> $DIR/equality2.rs:25:18
    |
 LL | fn hide<T: Foo>(x: T) -> impl Foo {
-   |                          -------- this is the found opaque type
+   |                          -------- the found opaque type
 ...
 LL |     let _: u32 = hide(0_u32);
    |            ---   ^^^^^^^^^^^ expected `u32`, found opaque type
@@ -16,7 +16,7 @@ error[E0308]: mismatched types
   --> $DIR/equality2.rs:31:18
    |
 LL | fn hide<T: Foo>(x: T) -> impl Foo {
-   |                          -------- this is the found opaque type
+   |                          -------- the found opaque type
 ...
 LL |     let _: i32 = Leak::leak(hide(0_i32));
    |            ---   ^^^^^^^^^^^^^^^^^^^^^^^ expected `i32`, found associated type
@@ -32,7 +32,10 @@ error[E0308]: mismatched types
   --> $DIR/equality2.rs:38:10
    |
 LL | fn hide<T: Foo>(x: T) -> impl Foo {
-   |                          -------- this is the expected opaque type
+   |                          --------
+   |                          |
+   |                          the expected opaque type
+   |                          the found opaque type
 ...
 LL |     x = (x.1,
    |          ^^^ expected `u32`, found `i32`
@@ -44,7 +47,10 @@ error[E0308]: mismatched types
   --> $DIR/equality2.rs:41:10
    |
 LL | fn hide<T: Foo>(x: T) -> impl Foo {
-   |                          -------- this is the expected opaque type
+   |                          --------
+   |                          |
+   |                          the expected opaque type
+   |                          the found opaque type
 ...
 LL |          x.0);
    |          ^^^ expected `i32`, found `u32`

--- a/src/test/ui/impl-trait/equality2.stderr
+++ b/src/test/ui/impl-trait/equality2.stderr
@@ -1,6 +1,9 @@
 error[E0308]: mismatched types
   --> $DIR/equality2.rs:25:18
    |
+LL | fn hide<T: Foo>(x: T) -> impl Foo {
+   |                          -------- this is the found opaque type
+...
 LL |     let _: u32 = hide(0_u32);
    |            ---   ^^^^^^^^^^^ expected `u32`, found opaque type
    |            |
@@ -12,6 +15,9 @@ LL |     let _: u32 = hide(0_u32);
 error[E0308]: mismatched types
   --> $DIR/equality2.rs:31:18
    |
+LL | fn hide<T: Foo>(x: T) -> impl Foo {
+   |                          -------- this is the found opaque type
+...
 LL |     let _: i32 = Leak::leak(hide(0_i32));
    |            ---   ^^^^^^^^^^^^^^^^^^^^^^^ expected `i32`, found associated type
    |            |
@@ -25,6 +31,9 @@ LL |     let _: i32 = Leak::leak(hide(0_i32));
 error[E0308]: mismatched types
   --> $DIR/equality2.rs:38:10
    |
+LL | fn hide<T: Foo>(x: T) -> impl Foo {
+   |                          -------- this is the expected opaque type
+...
 LL |     x = (x.1,
    |          ^^^ expected `u32`, found `i32`
    |
@@ -34,6 +43,9 @@ LL |     x = (x.1,
 error[E0308]: mismatched types
   --> $DIR/equality2.rs:41:10
    |
+LL | fn hide<T: Foo>(x: T) -> impl Foo {
+   |                          -------- this is the expected opaque type
+...
 LL |          x.0);
    |          ^^^ expected `i32`, found `u32`
    |

--- a/src/test/ui/issues/issue-24036.stderr
+++ b/src/test/ui/issues/issue-24036.stderr
@@ -1,6 +1,8 @@
 error[E0308]: mismatched types
   --> $DIR/issue-24036.rs:3:9
    |
+LL |     let mut x = |c| c + 1;
+   |                 --------- this is the expected closure
 LL |     x = |c| c + 1;
    |         ^^^^^^^^^ expected closure, found a different closure
    |

--- a/src/test/ui/issues/issue-24036.stderr
+++ b/src/test/ui/issues/issue-24036.stderr
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
   --> $DIR/issue-24036.rs:3:9
    |
 LL |     let mut x = |c| c + 1;
-   |                 --------- this is the expected closure
+   |                 --------- the expected closure
 LL |     x = |c| c + 1;
    |         ^^^^^^^^^ expected closure, found a different closure
    |

--- a/src/test/ui/kindck/kindck-nonsendable-1.stderr
+++ b/src/test/ui/kindck/kindck-nonsendable-1.stderr
@@ -5,7 +5,9 @@ LL | fn bar<F:FnOnce() + Send>(_: F) { }
    |    ---              ---- required by this bound in `bar`
 ...
 LL |     bar(move|| foo(x));
-   |     ^^^ `std::rc::Rc<usize>` cannot be sent between threads safely
+   |     ^^^ ------------- within this `[closure@$DIR/kindck-nonsendable-1.rs:9:9: 9:22 x:std::rc::Rc<usize>]`
+   |     |
+   |     `std::rc::Rc<usize>` cannot be sent between threads safely
    |
    = help: within `[closure@$DIR/kindck-nonsendable-1.rs:9:9: 9:22 x:std::rc::Rc<usize>]`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<usize>`
    = note: required because it appears within the type `[closure@$DIR/kindck-nonsendable-1.rs:9:9: 9:22 x:std::rc::Rc<usize>]`

--- a/src/test/ui/no-send-res-ports.stderr
+++ b/src/test/ui/no-send-res-ports.stderr
@@ -1,13 +1,20 @@
 error[E0277]: `std::rc::Rc<()>` cannot be sent between threads safely
   --> $DIR/no-send-res-ports.rs:29:5
    |
-LL |     thread::spawn(move|| {
-   |     ^^^^^^^^^^^^^ `std::rc::Rc<()>` cannot be sent between threads safely
+LL |       thread::spawn(move|| {
+   |  _____^^^^^^^^^^^^^_-
+   | |     |
+   | |     `std::rc::Rc<()>` cannot be sent between threads safely
+LL | |
+LL | |         let y = x;
+LL | |         println!("{:?}", y);
+LL | |     });
+   | |_____- within this `[closure@$DIR/no-send-res-ports.rs:29:19: 33:6 x:main::Foo]`
    | 
   ::: $SRC_DIR/libstd/thread/mod.rs:LL:COL
    |
-LL |     F: Send + 'static,
-   |        ---- required by this bound in `std::thread::spawn`
+LL |       F: Send + 'static,
+   |          ---- required by this bound in `std::thread::spawn`
    |
    = help: within `[closure@$DIR/no-send-res-ports.rs:29:19: 33:6 x:main::Foo]`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<()>`
    = note: required because it appears within the type `Port<()>`

--- a/src/test/ui/not-clone-closure.stderr
+++ b/src/test/ui/not-clone-closure.stderr
@@ -1,8 +1,14 @@
 error[E0277]: the trait bound `S: std::clone::Clone` is not satisfied in `[closure@$DIR/not-clone-closure.rs:7:17: 9:6 a:S]`
   --> $DIR/not-clone-closure.rs:11:23
    |
-LL |     let hello = hello.clone();
-   |                       ^^^^^ within `[closure@$DIR/not-clone-closure.rs:7:17: 9:6 a:S]`, the trait `std::clone::Clone` is not implemented for `S`
+LL |       let hello = move || {
+   |  _________________-
+LL | |         println!("Hello {}", a.0);
+LL | |     };
+   | |_____- within this `[closure@$DIR/not-clone-closure.rs:7:17: 9:6 a:S]`
+LL | 
+LL |       let hello = hello.clone();
+   |                         ^^^^^ within `[closure@$DIR/not-clone-closure.rs:7:17: 9:6 a:S]`, the trait `std::clone::Clone` is not implemented for `S`
    |
    = note: required because it appears within the type `[closure@$DIR/not-clone-closure.rs:7:17: 9:6 a:S]`
 

--- a/src/test/ui/suggestions/fn-or-tuple-struct-without-args.stderr
+++ b/src/test/ui/suggestions/fn-or-tuple-struct-without-args.stderr
@@ -236,7 +236,7 @@ error[E0308]: mismatched types
   --> $DIR/fn-or-tuple-struct-without-args.rs:46:20
    |
 LL |     let closure = || 42;
-   |                   -- closure defined here
+   |                   ----- this is the found closure
 LL |     let _: usize = closure;
    |            -----   ^^^^^^^
    |            |       |

--- a/src/test/ui/suggestions/fn-or-tuple-struct-without-args.stderr
+++ b/src/test/ui/suggestions/fn-or-tuple-struct-without-args.stderr
@@ -236,7 +236,7 @@ error[E0308]: mismatched types
   --> $DIR/fn-or-tuple-struct-without-args.rs:46:20
    |
 LL |     let closure = || 42;
-   |                   ----- this is the found closure
+   |                   ----- the found closure
 LL |     let _: usize = closure;
    |            -----   ^^^^^^^
    |            |       |

--- a/src/test/ui/suggestions/opaque-type-error.stderr
+++ b/src/test/ui/suggestions/opaque-type-error.stderr
@@ -2,7 +2,7 @@ error[E0308]: `if` and `else` have incompatible types
   --> $DIR/opaque-type-error.rs:20:9
    |
 LL |   fn thing_two() -> impl Future<Output = Result<(), ()>> {
-   |                     ------------------------------------ this is the found opaque type
+   |                     ------------------------------------ the found opaque type
 ...
 LL | /     if true {
 LL | |         thing_one()

--- a/src/test/ui/suggestions/opaque-type-error.stderr
+++ b/src/test/ui/suggestions/opaque-type-error.stderr
@@ -1,6 +1,9 @@
 error[E0308]: `if` and `else` have incompatible types
   --> $DIR/opaque-type-error.rs:20:9
    |
+LL |   fn thing_two() -> impl Future<Output = Result<(), ()>> {
+   |                     ------------------------------------ this is the found opaque type
+...
 LL | /     if true {
 LL | |         thing_one()
    | |         ----------- expected because of this

--- a/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.nll.stderr
+++ b/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.nll.stderr
@@ -13,7 +13,7 @@ LL |     let z: i32 = x;
    |            expected due to this
 ...
 LL | type WrongGeneric<T> = impl 'static;
-   | ------------------------------------ this is the found opaque type
+   | ------------------------------------ the found opaque type
    |
    = note:     expected type `i32`
            found opaque type `WrongGeneric::<&{integer}>`

--- a/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.nll.stderr
+++ b/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.nll.stderr
@@ -11,6 +11,9 @@ LL |     let z: i32 = x;
    |            ---   ^ expected `i32`, found opaque type
    |            |
    |            expected due to this
+...
+LL | type WrongGeneric<T> = impl 'static;
+   | ------------------------------------ this is the found opaque type
    |
    = note:     expected type `i32`
            found opaque type `WrongGeneric::<&{integer}>`

--- a/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.stderr
+++ b/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.stderr
@@ -13,7 +13,7 @@ LL |     let z: i32 = x;
    |            expected due to this
 ...
 LL | type WrongGeneric<T> = impl 'static;
-   | ------------------------------------ this is the found opaque type
+   | ------------------------------------ the found opaque type
    |
    = note:     expected type `i32`
            found opaque type `WrongGeneric::<&{integer}>`

--- a/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.stderr
+++ b/src/test/ui/type-alias-impl-trait/generic_type_does_not_live_long_enough.stderr
@@ -11,6 +11,9 @@ LL |     let z: i32 = x;
    |            ---   ^ expected `i32`, found opaque type
    |            |
    |            expected due to this
+...
+LL | type WrongGeneric<T> = impl 'static;
+   | ------------------------------------ this is the found opaque type
    |
    = note:     expected type `i32`
            found opaque type `WrongGeneric::<&{integer}>`

--- a/src/test/ui/type-alias-impl-trait/never_reveal_concrete_type.stderr
+++ b/src/test/ui/type-alias-impl-trait/never_reveal_concrete_type.stderr
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
   --> $DIR/never_reveal_concrete_type.rs:13:27
    |
 LL | type NoReveal = impl std::fmt::Debug;
-   | ------------------------------------- this is the found opaque type
+   | ------------------------------------- the found opaque type
 ...
 LL |     let _: &'static str = x;
    |            ------------   ^ expected `&str`, found opaque type

--- a/src/test/ui/type-alias-impl-trait/never_reveal_concrete_type.stderr
+++ b/src/test/ui/type-alias-impl-trait/never_reveal_concrete_type.stderr
@@ -1,6 +1,9 @@
 error[E0308]: mismatched types
   --> $DIR/never_reveal_concrete_type.rs:13:27
    |
+LL | type NoReveal = impl std::fmt::Debug;
+   | ------------------------------------- this is the found opaque type
+...
 LL |     let _: &'static str = x;
    |            ------------   ^ expected `&str`, found opaque type
    |            |

--- a/src/test/ui/type-alias-impl-trait/no_revealing_outside_defining_module.stderr
+++ b/src/test/ui/type-alias-impl-trait/no_revealing_outside_defining_module.stderr
@@ -1,6 +1,9 @@
 error[E0308]: mismatched types
   --> $DIR/no_revealing_outside_defining_module.rs:15:19
    |
+LL |     pub type Boo = impl ::std::fmt::Debug;
+   |     -------------------------------------- this is the found opaque type
+...
 LL |     let _: &str = bomp();
    |            ----   ^^^^^^ expected `&str`, found opaque type
    |            |
@@ -12,6 +15,9 @@ LL |     let _: &str = bomp();
 error[E0308]: mismatched types
   --> $DIR/no_revealing_outside_defining_module.rs:19:5
    |
+LL |     pub type Boo = impl ::std::fmt::Debug;
+   |     -------------------------------------- this is the expected opaque type
+...
 LL | fn bomp() -> boo::Boo {
    |              -------- expected `Boo` because of return type
 LL |     ""

--- a/src/test/ui/type-alias-impl-trait/no_revealing_outside_defining_module.stderr
+++ b/src/test/ui/type-alias-impl-trait/no_revealing_outside_defining_module.stderr
@@ -2,7 +2,7 @@ error[E0308]: mismatched types
   --> $DIR/no_revealing_outside_defining_module.rs:15:19
    |
 LL |     pub type Boo = impl ::std::fmt::Debug;
-   |     -------------------------------------- this is the found opaque type
+   |     -------------------------------------- the found opaque type
 ...
 LL |     let _: &str = bomp();
    |            ----   ^^^^^^ expected `&str`, found opaque type
@@ -16,7 +16,7 @@ error[E0308]: mismatched types
   --> $DIR/no_revealing_outside_defining_module.rs:19:5
    |
 LL |     pub type Boo = impl ::std::fmt::Debug;
-   |     -------------------------------------- this is the expected opaque type
+   |     -------------------------------------- the expected opaque type
 ...
 LL | fn bomp() -> boo::Boo {
    |              -------- expected `Boo` because of return type

